source/common/common/bit_array.h

```diff
@@ -0,0 +1,137 @@
+   *  to represent 10 distinct values we need at least 4 bits as the width.
+   * @param num_items the number of elements the bit array must hold.
+   */
+  BitArray(int width, size_t num_items)
+      : array_start_(std::make_unique<uint8_t[]>(bytesNeeded(width, num_items))),
+        end_(array_start_.get() + bytesNeeded(width, num_items)), bit_width_(width),
+        // This will fit in a uint32_t as with the maximum shift of 32, we'd
+        // subtract one to fit in 32 bits.
+        mask_(static_cast<uint32_t>((static_cast<uint64_t>(1) << width) - 1)),
+        num_items_(num_items) {
+    RELEASE_ASSERT(width <= MaxBitWidth, "Using BitArray with invalid parameters.");
+    static_assert(
+        sizeof(uint8_t*) == 8,
+        "Pointer underlying size is not 8 bytes, are we running on a 64-bit architecture?");
+  }
+
+  static const int MaxBitWidth = 32;
+  static const int WordSize = sizeof(uint8_t*);
+
+  /**
+   * Gets the fixed width bit at the given index.
+   * Assumes the the provided index is in bounds.
+   */
+  inline uint32_t get(size_t index) const {
+    RELEASE_ASSERT(index < num_items_, "BitArray requested index out of bounds");
+    const size_t bit0_offset = index * bit_width_;
+    const size_t byte0_offset = bit0_offset >> 3;
+    const uint8_t* byte0 = array_start_.get() + byte0_offset;
+    const size_t index_of_0th_bit = bit0_offset & 0x7;
+    // It is not in the last word, we can memcpy.
+    if (byte0 + WordSize <= end_) {
+      return (LoadUnsignedWord(byte0) >> index_of_0th_bit) & mask_;
+    }
+    // We're at the last word, and we've allocated at least a word of memory.
+    const uint8_t* last_word = end_ - WordSize;
+    return LoadUnsignedWord(last_word) >> (index_of_0th_bit + ((byte0 - last_word) << 3)) & mask_;
+  }
+
+  /**
+   * Sets the given fixed width bit to the given value.
+   */
+  inline void set(size_t index, uint32_t value) {
+    if (index > num_items_) {
+      ENVOY_BUG(true, "BitArray::set requested index out of bounds");
+      return;
+    }
+    const size_t bit0_offset = index * bit_width_;
+    const size_t byte0_offset = bit0_offset >> 3;
+    uint8_t* byte0 = array_start_.get() + byte0_offset;
+
+    // We are not at the end.
+    if (byte0 + WordSize <= end_) {
+      const size_t index_of_0th_bit = bit0_offset & 0x7;
+      const uint64_t shifted_value = static_cast<uint64_t>(value) << index_of_0th_bit;
+
+      const uint64_t mask_to_write = ((static_cast<uint64_t>(1) << bit_width_) - 1)
+                                     << index_of_0th_bit;
+      const uint64_t mask_to_preserve = ~mask_to_write;
+      const uint64_t value_to_store =
+          ((LoadUnsignedWord(byte0) & mask_to_preserve) | (shifted_value & mask_to_write));
+      StoreUnsignedWord(byte0, value_to_store);
+    } else {
+      // We're at the end, so we'll be starting from the last word
+      uint8_t* last_word = const_cast<uint8_t*>(end_) - WordSize;
```

Comment:
What do you think about padding the allocated space by `WordSize` so we don't need the else code block? Either way is fine with me, but doing so will make the code easier to follow.

Also, could you please add more comments? The code here is very hard for me to follow/reason about.

Reply:
done.